`lib.fileset`: Representation for empty file sets without a base path (#257351)
NixOS/nixpkgs: lib/fileset/README.md

- `_type` (constant string `"fileset"`):
  Tag to indicate this value is a file set.

- `_internalVersion` (constant `3`, the current version):
  Version of the representation.

> - `_internalIsEmptySingleton` (bool):
>   Whether this file set is the singleton representing an empty file set.

Reviewer: ```suggestion
- `_internalIsEmpty` (bool):
  Whether this file set is the empty file set.
```

Not sure if it's required to call it a singleton in the parameter. It's just *the* root-less empty set, regardless of how it's implemented.

Author: I wanted to avoid calling it that because there's multiple ways to represent an empty fileset: `null` as the tree is another representation, `{ }` is another and there's many more if you have some files in the base path.

Reviewer: What about this then?

```suggestion
- `_internalIsEmptyRootless` (bool):
  Whether this file set is the empty file set without a base path.
```

Author: I like that, should be `_intetrnalIsEmptyBaseless` though. Will change

Author: Although, it's not _just_ without a base path, it's also without a file set tree..

Reviewer: Is it even possible to have a file set with a tree but without a base?

Also why isn't this property returned by a function? While it's all internal and users should not mess with it, the new field introduces another moving part and opportunity for error.

Author: At least currently no base path also means no tree and the other way around, so yeah it might be okay to word it like that. In the end it doesn't matter much since indeed it is internal.

> Also why isn't this property returned by a function? While it's all internal and users should not mess with it, the new field introduces another moving part and opportunity for error.

It's not relevant for the user, it's an implementation detail. To the user it behaves just like they would expect. Exposing this as a function would expose implementation details.

Reviewer: To be clear, that helper function would also be internal.

Reviewer 2: Besides being returned for an empty union, isn't its unique property that its the identity of any union operation?
You may consider `_isUnionIdentity`.

- `_type` (constant string `"fileset"`):
  Tag to indicate this value is a file set.

> - `_internalVersion` (constant `3`, the current version):

Reviewer 2: Should this be in a file like `fileset/internal/README.md`?

Author: I view `lib/fileset/README.md` as the internal contributor documentation, where I think this makes sense to have. It's not part of the public interface, but you need to know about it to contribute, which can mean to change the internals

Reviewer 2: It should link to the user docs.
